kythe/java/com/google/devtools/kythe/platform/kzip/KZipReader.java

```diff
@@ -59,19 +61,42 @@ private static String getRootPrefix(ZipFile zipFile) {
     }
     String rootPrefix = root.getName();
 
+    Set<String> jsonUnits = new HashSet<>();
+    Set<String> protoUnits = new HashSet<>();
+    String jsonPrefix = rootPrefix + KZip.Encoding.JSON.getSubdirectory();
+    String protoPrefix = rootPrefix + KZip.Encoding.PROTO.getSubdirectory();
+
     // Make sure each path in the kzip has the same root.
+    // Also accumulate potential unit digests.
     while (zipEntries.hasMoreElements()) {
       ZipEntry zipEntry = zipEntries.nextElement();
-      if (!zipEntry.getName().startsWith(rootPrefix)) {
-        throw new KZipException("Invalid entry (bad root): " + zipEntry.getName());
+      String name = zipEntry.getName();
+      if (!name.startsWith(rootPrefix)) {
+        throw new KZipException("Invalid entry (bad root): " + name);
+      }
+      if (name.startsWith(jsonPrefix)) {
```

Comment:
should this be name.startsWith(jsonPrefix + "/") to prevent unexpected directory structures?

Reply:
Done